### PR TITLE
py-hail: new package

### PR DESCRIPTION
**Pull request by teaguesterling**
Adding the `py-hail` package to build and install `HAIL` python libraries and all upstream C, Java, and Scala libraries.

This depends on a number of other PRs, most of which have already been merged. The final two are:
 - #45478 
 - #45477 

This builds the HAIL package for genomic analysis to be used on a Spark cluster or locally.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py:33]: Is there any reasonable way to depend on `catch2`?

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py]: Can you add a brief comment above the `patch` directive explaining:
1. What is the effect of the patch
2. Why it is needed
3. Where it was taken from

? In particular it's not clear to me why using `"` shouldn't work, since `<` is more restrictive. Are we injecting some wrong `-I` statements?

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py]: :thinking:

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py]: Can we define a `flag_handler` instead?

**Review by alalazo — COMMENTED** [review]
A few initial comments.

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py:33]: > Is there any reasonable way to depend on `catch2`?

They version it into the source tree during the installation, which seemed difficult to address with spack. The build process is rather quirky, to say the least.

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py]: Actually, we may be able to drop this. Was added in a very early attempt.

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py]: There's probably some environmental variables that can be set for optimal performance. Not required but perhaps for a future iteration. Will remove for now.

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py]: Removed

**Comment by teaguesterling** [comment]
> A few initial comments.

Addressed the comments but would like to do another test build locally to be sure before merging

File: var/spack/repos/builtin/packages/py-hail/fix-lz4-import-builtins.patch
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/fix-lz4-import-builtins.patch:1]: Patch file was not removed

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py]: What does "native" mean in this context? If this is related to `-march=` kind of flags, can we avoid the variant and inspect the `spec.target`?

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py:89]: How comes these three deps are `build, run` instead of `build, link` ?

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-hail/package.py]: Some typos in here :slightly_smiling_face:

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py]: It's a naming thing. There's a way to build the HAIL with some prepackaged shared libraries, IIRC. Maybe worth removing.

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py:89]: Whoops!

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py]: :facepalm:

File: var/spack/repos/builtin/packages/py-hail/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/py-hail/package.py]: Updated the description (and also refined to dependencies around this). Tested both builds.

**Comment by teaguesterling** [comment]
@alalazo Thanks for taking the time to review this one. Should now be good to go assuming the CI agrees.

**Review by alalazo — APPROVED** [review]

**Pull request merged**
